src/Conversion/ONNXToTOSA/Math/Gemm.cpp

```diff
@@ -0,0 +1,208 @@
+//
+// This file lowers ONNX Gemm operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "llvm/ADT/SmallVector.h"
+#include <cstdint>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+
+
+class ONNXGemmOpLoweringToTOSA : public OpConversionPattern<ONNXGemmOp> {
+public:
+  using OpConversionPattern<ONNXGemmOp>::OpConversionPattern;
+  LogicalResult rewriteToTosaFC(ONNXGemmOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const {
+    Value A = adaptor.A();
+    Value B = adaptor.B();
+    Value C = adaptor.C();
+
+    Type resultType = getTypeConverter()->convertType(op.getResult().getType());
+
+    // If no bias is given to the GEMM operator, we create a 1D bias with all
+    // zeroes
+    if (C.getType().isa<mlir::NoneType>()) {
+      // B is supposed to have 
+      ArrayRef<int64_t> cformat(B.getType().cast<TensorType>().getShape()[1]);
+      std::vector<float> elements = {};
+      for (int i = 0; i < cformat[0]; ++i)
+        elements.push_back(0.0F);
+      C = mlir::tosa::getConstTensor<float>(rewriter, op, elements, cformat).value();
+    }
+
+    rewriter.replaceOpWithNewOp<tosa::FullyConnectedOp>(op, resultType, A, B, C);
+    return success();
+  }
+
+  /// The GEMM can be described as a FullyConnected operator
+  /// Y = AB^T + C if we perform a transpose on B only with
+  /// alpha and beta factors set to 1.
+  /// The constraints on rank were taken from the torch implementation check
+  /// Input A must be of rank 2 (input)
+  /// Input B must be of rank 2 (weights)
+  /// Input C must be of rank 1 (bias)
+  static bool checkLegalTosaFCOp(ONNXGemmOp op, OpAdaptor adaptor) {
+    Value A = op.A();
+    Value B = op.B();
+    Value C = op.C();
+
+    auto AType = A.getType().cast<RankedTensorType>();
+    auto BType = B.getType().cast<RankedTensorType>();
+
+    // If C is present, it can only be of rank 1
+    if (C.getType().isa<RankedTensorType>() &&
+        C.getType().cast<RankedTensorType>().getRank() != 1) {
+      return false;
+    }
+    // Input tensor must be of rank 2
+    // Weights must also be of rank 2
+    if (AType.getRank() != 2 || BType.getRank() != 2) {
+      return false;
+    }
+    // Both alpha and beta must be 1
+    if ((adaptor.alpha().convertToFloat() != 1.0F) ||
+        (adaptor.beta().convertToFloat() != 1.0F)) {
+      return false;
+    }
+    // Only Transpose B must be enabled
+    if (adaptor.transA() != 0 || adaptor.transB() != 1) {
+      return false;
+    }
+    return true;
+  }
+  LogicalResult matchAndRewrite(ONNXGemmOp op, OpAdaptor adaptor,
```

Comment:
I'd prefer this to be the 1st function defined in the class so it is easier for find.